internal/controller/dataimportjob_controller.go

```diff
@@ -0,0 +1,427 @@
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	everestv1alpha1 "github.com/percona/everest-operator/api/v1alpha1"
+	"github.com/percona/everest-operator/api/v1alpha1/dataimporterspec"
+)
+
+const (
+	dataImporterRequestSecretNameSuffix = "-data-import-request"
+	dataImportJSONSecretKey             = "request.json"
+)
+
+type DataImportJobReconciler struct {
+	Client client.Client
+	Scheme *runtime.Scheme
+}
+
+func (r *DataImportJobReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("DataImportJob").
+		For(&everestv1alpha1.DataImportJob{}).
+		Owns(&batchv1.Job{}).
+		Owns(&corev1.Secret{}).
+		Complete(r)
+}
+
+//+kubebuilder:rbac:groups=everest.percona.com,resources=dataimportjobs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=everest.percona.com,resources=dataimportjobs/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=everest.percona.com,resources=dataimportjobs/finalizers,verbs=update
+//+kubebuilder:rbac:groups=batch,resources=jobs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=core,resources=secrets,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=everest.percona.com,resources=dataimporters,verbs=get;list;watch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.1/pkg/reconcile
+func (r *DataImportJobReconciler) Reconcile(
+	ctx context.Context,
+	req ctrl.Request,
+) (rr ctrl.Result, rerr error) {
+	diJob := &everestv1alpha1.DataImportJob{}
+	if err := r.Client.Get(ctx, req.NamespacedName, diJob); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	l := log.FromContext(ctx)
+	l.Info("Reconciling", "name", diJob.GetName())
+
+	if diJob.Status.Phase == everestv1alpha1.DataImportJobPhaseCompleted {
+		// Already complete, no need to reconcile again.
+		return ctrl.Result{}, nil
+	}
+
+	// Reset the status, we will build a new one by observing the current state on each reconcile.
+	startedAt := diJob.Status.StartedAt
+	diJob.Status = everestv1alpha1.DataImportJobStatus{}
+	diJob.Status.LastObservedGeneration = diJob.GetGeneration()
+	if startedAt != nil && !startedAt.Time.IsZero() {
+		diJob.Status.StartedAt = startedAt
+	}
+
+	// Sync status on finishing reconciliation.
+	defer func() {
+		if updErr := r.Client.Status().Update(ctx, diJob); updErr != nil {
+			l.Error(updErr, "Failed to update data import job status")
+			rerr = errors.Join(rerr, updErr)
+		}
+	}()
+
+	// Get the referenced data importer.
+	di := &everestv1alpha1.DataImporter{}
+	if err := r.Client.Get(ctx, client.ObjectKey{
+		Name: diJob.Spec.DataImporterName,
+	}, di); err != nil {
+		diJob.Status.Phase = everestv1alpha1.DataImportJobPhaseError
+		diJob.Status.Message = err.Error()
```

Comment:
Why not add a wrapper message like we did in the other errors below?

Reply:
updated